sql/analyzer/optimization_rules.go

```diff
@@ -149,21 +149,40 @@ func moveJoinConditionsToFilter(ctx *sql.Context, a *Analyzer, n sql.Node, scope
 		return node, transform.SameTree, nil
 	}
 
-	return transform.NodeWithCtx(node, func(transform.Context) bool { return true }, func(c transform.Context) (sql.Node, transform.TreeIdentity, error) {
-		if c.Node != topJoin {
-			return c.Node, transform.SameTree, nil
+	resultNode, resultIdentity, err := transform.Node(node, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		children := n.Children()
+		if children == nil || len(children) == 0 {
+			return n, transform.SameTree, nil
+		}
+		isParentOfTopJoin := false
+		for _, child := range children {
+			if child == topJoin {
+				isParentOfTopJoin = true
+				break
+			}
 		}
-		switch n := c.Parent.(type) {
+		if !isParentOfTopJoin {
+			return n, transform.SameTree, nil
+		}
+
+		switch imp := n.(type) {
```

Comment:
```suggestion
		switch n := n.(type) {
```

Reply:
Feels really strange to reassign `n` like this, but the scoping seems to work out, so going with this approach.